samples/controller/controller.go

```diff
@@ -69,189 +66,54 @@ type Controller struct {
 	sampleclientset clientset.Interface
 
 	deploymentsLister appslisters.DeploymentLister
-	deploymentsSynced cache.InformerSynced
 	foosLister        listers.FooLister
-	foosSynced        cache.InformerSynced
-
-	// workqueue is a rate limited work queue. This is used to queue work to be
-	// processed instead of performing it as soon as a change happens. This
-	// means we can ensure we only process a fixed amount of resources at a
-	// time, and makes it easy to ensure we are never processing the same item
-	// simultaneously in two different workers.
-	workqueue workqueue.RateLimitingInterface
+
 	// recorder is an event recorder for recording Event resources to the
 	// Kubernetes API.
 	recorder record.EventRecorder
 }
 
+var MapToDeploymentToTypes = []metav1.GroupVersionKind{
+	{Group: "samplecontroller.k8s.io", Version: "v1alpha1", Kind: "Foo"},
+}
+
 // NewController returns a new sample controller
-func NewController(
-	kubeclientset kubernetes.Interface,
-	sampleclientset clientset.Interface,
-	kubeInformerFactory kubeinformers.SharedInformerFactory,
-	sampleInformerFactory informers.SharedInformerFactory) *Controller {
-
-	// obtain references to shared index informers for the Deployment and Foo
-	// types.
-	deploymentInformer := kubeInformerFactory.Apps().V1().Deployments()
-	fooInformer := sampleInformerFactory.Samplecontroller().V1alpha1().Foos()
-
-	// Create event broadcaster
-	// Add sample-controller types to the default Kubernetes Scheme so Events can be
-	// logged for sample-controller types.
+func NewController(iargs args.InjectArgs) *controller.GenericController {
 	samplescheme.AddToScheme(scheme.Scheme)
-	glog.V(4).Info("Creating event broadcaster")
-	eventBroadcaster := record.NewBroadcaster()
-	eventBroadcaster.StartLogging(glog.Infof)
-	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
-	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
-
-	controller := &Controller{
-		kubeclientset:     kubeclientset,
-		sampleclientset:   sampleclientset,
-		deploymentsLister: deploymentInformer.Lister(),
-		deploymentsSynced: deploymentInformer.Informer().HasSynced,
-		foosLister:        fooInformer.Lister(),
-		foosSynced:        fooInformer.Informer().HasSynced,
-		workqueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Foos"),
-		recorder:          recorder,
+
+	c := &Controller{
+		kubeclientset:     iargs.KubernetesClientSet,
+		sampleclientset:   iargs.Clientset,
+		deploymentsLister: iargs.KubernetesInformers.Apps().V1().Deployments().Lister(),
+		foosLister:        iargs.Informers.Samplecontroller().V1alpha1().Foos().Lister(),
+		recorder:          iargs.CreateRecorder(controllerAgentName),
+	}
```

Comment:
wondering if Controller struct can have iargs in it ? because everything the controller needs is coming from iargs ?

Reply:
That'd be somewhat similar to inject cobra deep down everywhere IMHO. Objects like Controller shouldn't be coupled with iargs, or with the source of their parameters.

Reply:
Yeah, I think that might be the way to go.